WIP: provide COMMIT_MESSAGE as an env var

jeff-knurek opened the pull request:
## [JENKINS-52490](https://issues.jenkins-ci.org/browse/JENKINS-52490) - Plugin should provide GIT_COMMIT_MESSAGE as Environment Variable

The commit message should be available as an env var.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.adoc) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [ ] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [ ] No spotbugs warnings were introduced with my changes
- [ ] I have interactively tested my changes

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

vwin commented:
I think so ~

MarkEWaite commented on src/main/java/hudson/plugins/git/GitSCM.java:
```diff
@@ -1224,16 +1227,30 @@ public void checkout(Run<?, ?> build, Launcher launcher, FilePath workspace, Tas
+        return "";
     }
 
+
```
Please don't add extra white space in a change.  The white space is already messy enough in the plugin without adding duplicated empty lines
```suggestion
```

jeff-knurek commented:
I've updated with the documentation bits that seem to be needed. I've also added a test, but the test doesn't actually pass. Spent some time trying to debug it, but may need to take a different approach.

otradovec commented:
@jeff-knurek  I don't understand why are you asserting String with Env object. Maybe you wanted 
`assertEquals("Commit message should be an env var", title, getEnvVars(p).get("some_key","default"));`
But I am a newbie, so I am missing something propably.

jeff-knurek commented:
@otradovec you're totally right. I had pushed some in-progress code from while I was debugging the test. I've now pushed what I expect to work (but doesn't)

otradovec commented:
What about using GIT_COMMIT instead? I don't see GIT_COMMIT_TITLE here https://javadoc.jenkins.io/plugin/git/hudson/plugins/git/GitSCM.html
Or is that defined somewhere else?

otradovec commented:
Btw this: https://javadoc.jenkins-ci.org/hudson/model/Run.html#getEnvVars--
getEnvVars is deprecated

jeff-knurek commented:
`GIT_COMMIT_TITLE` is what I'm adding in this PR

----
Your tip about `getEnvVars` being deprecated gave me an idea, but it still didn't work.
Because right now, the `getEnvVars` is actually this one, based off the builder, and not the run:
https://github.com/jenkinsci/git-plugin/blob/master/src/test/java/hudson/plugins/git/AbstractGitTestCase.java#L282
so I tried using the one from the run, but it also doesn't have the new Env Var :(

maneetgoyal commented:
Hi @MarkEWaite, is there anything remaining for this PR? It's a useful feature.

jeff-knurek commented:
I never got the unit tests to work. I think that's pretty critical to get this out. I'm not a Java/Jenkins developer though, so I think I've run out of things to try.

raghavgarg098 commented on src/test/java/hudson/plugins/git/GitSCMTest.java:
```diff
@@ -3064,6 +3064,24 @@ public void testCommitMessageIsPrintedToLogs() throws Exception {
     }
 
+    @Test
+    public void testCommitMessageIsEnvVar() throws Exception {
```
Similar to this I was trying to set environment variable for https://issues.jenkins.io/browse/JENKINS-38699. Does this test case run correctly in the current state?

raghavgarg098 commented on src/main/java/hudson/plugins/git/GitSCM.java:
```diff
@@ -1352,6 +1353,9 @@ public void checkout(Run<?, ?> build, Launcher launcher, FilePath workspace, Tas
         try {
+            String shortMessage = getCommitMessage(listener, git, revToBuild);
+            listener.getLogger().println("Commit message: \"" + shortMessage + "\"");
+            environment.put(GIT_COMMIT_TITLE, shortMessage);
```
@MarkEWaite In this(https://github.com/jenkinsci/git-plugin/pull/1049/files#r583842074) pull request you mentioned with this enviroment variable is set only for the process.
I want to understand what is the scope of adding a variable to EnvVars is? Since it is being added in git plugin would it be available to just the checkout step or it will be available in subsequent steps like build etc. ? From the bug description I believe we want to use this env variable to check if we want to run the build step based on the commit message. So would this env variable be available in the pipeline script in the build step with this ?

MarkEWaite commented:
Closing in favor of #1074
